execution/block-partitioner/src/v2/mod.rs

```diff
@@ -0,0 +1,719 @@
+// Copyright © Aptos Foundation
+
+#[cfg(test)]
+use crate::assert_deterministic_result;
+#[cfg(test)]
+use crate::test_utils::P2PBlockGenerator;
+use crate::{
+    get_anchor_shard_id,
+    v2::{conflicting_txn_tracker::ConflictingTxnTracker, counters::MISC_TIMERS_SECONDS},
+    BlockPartitioner, Sender,
+};
+use aptos_logger::info;
+use aptos_types::{
+    block_executor::partitioner::{
+        CrossShardDependencies, RoundId, ShardId, ShardedTxnIndex, SubBlock, SubBlocksForShard,
+        TransactionWithDependencies, TxnIndex,
+    },
+    state_store::state_key::StateKey,
+    transaction::analyzed_transaction::AnalyzedTransaction,
+};
+use dashmap::DashMap;
+#[cfg(test)]
+use rand::thread_rng;
+#[cfg(test)]
+use rand::Rng;
+use rayon::{
+    iter::ParallelIterator,
+    prelude::{IntoParallelIterator, IntoParallelRefIterator},
+    ThreadPool, ThreadPoolBuilder,
+};
+use serde::{Deserialize, Serialize};
+#[cfg(test)]
+use std::sync::Arc;
+use std::{
+    cmp,
+    collections::HashSet,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Mutex, RwLock,
+    },
+};
+
+/// The position of a txn in the block *before partitioning*.
+type OriginalTxnIdx = usize;
+
+/// Represent a specific storage location in a partitioning session.
+type StorageKeyIdx = usize;
+
+/// Represent a sender in a partitioning session.
+type SenderIdx = usize;
+
+/// Represent a txn after its position is finalized.
+/// Different from `aptos_types::block_executor::partitioner::ShardedTxnIndex`,
+#[derive(Copy, Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub struct ShardedTxnIndex2 {
+    pub round_id: RoundId,
+    pub shard_id: ShardId,
+    pub ori_txn_idx: OriginalTxnIdx,
+}
+
+impl Ord for ShardedTxnIndex2 {
+    fn cmp(&self, other: &Self) -> cmp::Ordering {
+        (self.round_id, self.shard_id, self.ori_txn_idx).cmp(&(
+            other.round_id,
+            other.shard_id,
+            other.ori_txn_idx,
+        ))
+    }
+}
+
+impl PartialOrd for ShardedTxnIndex2 {
+    fn partial_cmp(&self, other: &Self) -> Option<cmp::Ordering> {
+        (self.round_id, self.shard_id, self.ori_txn_idx).partial_cmp(&(
+            other.round_id,
+            other.shard_id,
+            other.ori_txn_idx,
+        ))
+    }
+}
+
+impl ShardedTxnIndex2 {
+    pub fn new(round_id: RoundId, shard_id: ShardId, pre_par_tid: OriginalTxnIdx) -> Self {
+        Self {
+            round_id,
+            shard_id,
+            ori_txn_idx: pre_par_tid,
+        }
+    }
+}
+
+mod conflicting_txn_tracker;
+mod counters;
+
+/// Basically `ShardedBlockPartitioner` but:
+/// - Not pre-partitioned by txn sender.
+/// - implemented more efficiently.
+pub struct PartitionerV2 {
+    thread_pool: ThreadPool,
```

Comment:
checkout my latest PR, we can use the `BlockPartitionerConfiguration` class for both V1 and V2.